idb/typeinf.py

```diff
@@ -0,0 +1,1713 @@
+
+        # args
+        if is_cm_cc_voidarg(self.cc):
+            return self
+        n = ts.dt()
+        if n > 256:
+            raise ValueError("invalid arg count!")
+        else:
+            for n in range(n):
+                arg = FuncArg()
+                arg.type = create_tinfo(til, ts.ref(), fields, fieldcmts)
+                if is_cm_cc_special_pe(self.cc):
+                    arg.argloc = self.deserialize_argloc(ts.get())
+                if ts.has_next() and ts.peek_u8() == FAH_BYTE:
+                    ts.seek(1)
+                    arg.flags = ts.de()
+                if fields is not None and n < len(fields):
+                    arg.name = fields[n]
+                if fieldcmts is not None and n < len(fieldcmts):
+                    arg.cmt = fieldcmts[n]
+                self.args.append(arg)
+        return self
+
+    def deserialize_argloc(self, type_string):
+        # TODO: deserialize_argloc
+        raise NotImplementedError("deserialize_argloc() not implemented.")
+
+
+# tattr_field Type attributes for udt fields
+TAFLD_BASECLASS = 0x0020  # field: do not include but inherit from the current field
+TAFLD_UNALIGNED = 0x0040  # field: unaligned field
+TAFLD_VIRTBASE = 0x0080  # field: virtual base (not supported yet)
+
+
+class UdtMember:
+    def __init__(self):
+        self.offset = 0
+        self.size = 0
+        self.name = None  # qstring
+        self.cmt = None  # qstring
+        self.type = None  # tinfo_t
+        self.effalign = 0
+        self.tafld_bits = 0
+        self.fda = 0
+
+    def is_unaligned(self):
+        return bool(self.tafld_bits & TAFLD_UNALIGNED)
+
+    def is_baseclass(self):
+        return bool(self.tafld_bits & TAFLD_BASECLASS)
+
+    def is_virtbase(self):
+        return bool(self.tafld_bits & TAFLD_VIRTBASE)
+
+
+# tattr_udt Type attributes for udts
+TAUDT_UNALIGNED = 0x0040  # struct: unaligned struct
+TAUDT_MSSTRUCT = 0x0020  # struct: gcc msstruct attribute
+TAUDT_CPPOBJ = 0x0080  # struct: a c++ object, not simple pod type
+
+
+class UdtTypeData(TypeData):
+    """An object to represent struct or union types"""
+
+    def __init__(self):
+        TypeData.__init__(self)
+        self.members = []
+        # total structure size in bytes
+        self.total_size = 0
+        # unpadded structure size in bytes
+        self.unpadded_size = 0
+        # effective structure alignment (in bytes)
+        self.effalign = 0
+        # TA... and TAUDT... bits.
+        self.taudt_bits = 0
+        # declared structure alignment (shift amount+1). 0 - unspecified
+        self.sda = 0
+        # pragma pack() alignment (shift amount)
+        self.pack = 0
+        self.is_union = False
+        self.ref = None
+
+    def deserialize(self, til, ts, fields, fieldcmts):
+        typ = ts.u8()
+        self.is_union = is_type_union(typ)
+        n = ts.complex_n()
+        if n == 0 and len(ts.rest()) >= 2 and ts.peek_u16() != 33009:
```

Comment:
Hmmm, where does 33009 come from? They handle this part in `deserialize_tinfo()`. It checks 3 things:
- Base type is `BT_COMPLEX`
- Flags is **not** `BTMT_TYPEDEF`
- The result from `get_complex_n()`, or `complex_n()`, in your case is 0.

Reply:
Yeah, I have encountered some situations where it is not a UDT reference, but N = 0,

So skip it with this feature.

Reply:
Maybe so
```
///< struct/union/enum/typedef.
///< format: <pre>
///   [dt N (N=field count) if !::BTMT_TYPEDEF]
///   if N == 0:
///     p_string name (unnamed types have names "anon_...")
///     [sdacl-typeattrs];
///   else, for struct & union:
///     if N == 0x7FFE   // Support for high (i.e., > 4095) 
members count
///       N = deserialize_de()
///     ALPOW = N & 0x7
///     MCNT = N >> 3
```